tensorflow/cc/gradients/nn_grad.cc

```diff
@@ -47,6 +48,38 @@ Status SoftmaxGrad(const Scope& scope, const Operation& op,
 }
 REGISTER_GRADIENT_OP("Softmax", SoftmaxGrad);
 
+Status SoftmaxCrossEntropyWithLogitsGrad(const Scope& scope,
+                                          const Operation& op,
+                                          const std::vector<Output>&
+                                          grad_inputs,
+                                          std::vector<Output>* grad_outputs) {
+  // Softmax gradient with cross entropy logits function
+  // We multiply the backprop for cost with the gradients - op.output[1]
+  // There is no gradient for labels
+
+  auto softmaxGrad = op.output(1);
+  auto gradLoss = grad_inputs[0];
+  auto gradGrad = grad_inputs[1];
+
+  auto tempGrad = Mul(scope, gradLoss, softmaxGrad);
+
+  // TODO: Check if this sufficient, need a check for null ?
+  if (gradGrad.op().output_type(0) != 0) {
+	  auto logits = op.input(0);
+	  auto softmax = ops::Softmax(scope, logits);
+	  auto prod = ops::MatMul(scope, gradGrad, softmax);
```

Comment:
In python `MatMul` is:
```python
math_ops.matmul(grad_grad[:, None, :],
                        softmax[:, :, None]), axis=1)
```
which I am not sure why it does? What is the effect of this?

Reply:
For ref the python code is:
```python
logits = op.inputs[0]
    softmax = nn_ops.softmax(logits)

    grad += ((grad_grad - array_ops.squeeze(
        math_ops.matmul(grad_grad[:, None, :],
                        softmax[:, :, None]), axis=1)) * softmax)
```